### DISCUSSION TITLE
brew pr-pull: Merge instead of close?

### DISCUSSION THREAD

**Discussion by skullydazed** [Homebrew/discussions#1842] [Tap maintenance and Homebrew development]
We recently setup `brew pr-pull` but during [one of the first runs](https://github.com/qmk/homebrew-qmk/pull/23) I noticed that instead of merging the PR it pushed a commit directly to master and then closed the PR. Is there a technical limitation behind this decision?

For us this is problematic on a number of fronts-

* It doesn't match reality- the PR was added to our repo but it looks like it wasn't
* It makes finding skipped versions harder
* It makes tracking a commit in `git log` back to a PR (and any relevant discussion) harder

Is it possible to change that behavior so that PRs are merged instead of being closed?

**Comment by carlocab** [comment]
This is somewhat of a technical limitation given the constraint that the bottle commit needs to be merged simultaneously with the formula changes.

You could maybe fix up test-bot so that it pushes the bottle commit to the PR branch, persuade test-bot that it doesn't need to do a rebuild for this new pushed commit, and then merge the PR from there.

You might need to fork test-bot for this -- not sure if we'd be able to adopt this for Homebrew/core, since contributors don't always give repo admins push access to their PR branch. Not sure if there's a setting in GitHub that forces PRs to be accessible to repo admins. GitHub also doesn't preserve commit signatures when rebasing using their UI.

**Reply by skullydazed** [reply]
Thanks for the response, I suspected there might be something like this going on. I don't know yet if I think the work here will be worth the effort but I'll give it a good look over.

> persuade test-bot that it doesn't need to do a rebuild for this new pushed commit

Would it be acceptable to accomplish this in a more general way by skipping builds for PRs that have bottles attached?

> You might need to fork test-bot for this -- not sure if we'd be able to adopt this for Homebrew/core, since contributors don't always give repo admins push access to their PR branch.

My thought around that is to make the behavior configurable. If we're getting ambitious it might be possible to fall-back to merging in the case where updating the PR fails. If it couldn't be accepted upstream even after that I would understand, too.

**Reply by carlocab** [reply]
> Would it be acceptable to accomplish this in a more general way by skipping builds for PRs that have bottles attached?

I'm not sure what you mean by this, though. How do you determine that a PR has a bottle attached? If by that you mean you check whether the PR modifies the `bottle` block: we do still regularly get PRs where users modify the `bottle` block even when they're not supposed to. Also, we haven't had a great experience with deciding whether certain workflows get run by parsing the diff, so I'm not optimistic about this approach.

**Reply by skullydazed** [reply]
Looks like I have a lot more to learn about how this process works, my assumptions are quickly breaking down. :)
